### PR TITLE
[Enhancement] JDBC connector supports TIME type

### LINKED ISSUE #24278: mysql catalog datatype `time` not support
<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
1. i have a mysql db whose table defination like this:
```
mysql> show create table dbgen_version;
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table         | Create Table                                                                                                                                                                                                                                                         |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| dbgen_version | CREATE TABLE `dbgen_version` (
  `dv_version` varchar(16) DEFAULT NULL,
  `dv_create_date` date DEFAULT NULL,
  `dv_create_time` time DEFAULT NULL,
  `dv_cmdline_args` varchar(200) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
and data in this table like this:
```
+------------+----------------+----------------+-----------------------------------------------------+
| dv_version | dv_create_date | dv_create_time | dv_cmdline_args                                     |
+------------+----------------+----------------+-----------------------------------------------------+
| 3.2.0      | 2023-05-25     | 20:06:25       | -SCALE 5 -DIR /opt/zkk/tpcds_data_5g/ -TERMINATE N  |
+------------+----------------+----------------+-----------------------------------------------------+
1 row in set (0.02 sec)
```
2. then i create a jdbc catalog by sql:
```
CREATE EXTERNAL CATALOG mysql_catalog
PROPERTIES
(
    "type"="jdbc",
    "user"="mysqluser",
    "password"="changeme",
    "jdbc_uri"="jdbc:mysql://localhost:3306",
    "driver_url"="file:/opt/starrocks/jdbc/mysql-connector-java-8.0.28.jar",
    "driver_class"="com.mysql.cj.jdbc.Driver"
);
```
4. when i query data from catalog error occurs:
```
mysql> select * from dbgen_version;
ERROR 1064 (HY000): Datatype of external table column [dv_create_time] is not supported!
```

### Expected behavior (Required)
```
+------------+----------------+----------------+-----------------------------------------------------+
| dv_version | dv_create_date | dv_create_time | dv_cmdline_args                                     |
+------------+----------------+----------------+-----------------------------------------------------+
| 3.2.0      | 2023-05-25     | 20:06:25       | -SCALE 5 -DIR /opt/zkk/tpcds_data_5g/ -TERMINATE N  |
+------------+----------------+----------------+-----------------------------------------------------+
1 row in set (0.02 sec)
```
### Real behavior (Required)
```
mysql> select * from dbgen_version;
ERROR 1064 (HY000): Datatype of external table column [dv_create_time] is not supported!
```

### StarRocks version (Required)
 - 3.0.0

### PR DESCRIPTION
**Pull request by JackWangCS**
Fixes #24278

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [x] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by stdpain — DISMISSED** [review]

**Comment by silverbullet233** [comment]
In addition to jdbc catalog, SR also supports the explicit creation of jdbc external tables to access data(ref:https://docs.starrocks.io/en-us/latest/using_starrocks/External_table#external-table-for-a-jdbc-compatible-database). We need to ensure that the TIME type also works in the latter way.

SR does not support the TIME type in the storage layer. If you specify a column of the TIME type when creating a table, an error will be reported.

However, the query of jdbc external table does not involve the storage layer. It may need special processing to ensure that the TIME type can be specified when creating the jdbc external table.

You can look at the logic related to table creation to ensure that what I said can work. @JackWangCS

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 5 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/jdbc_scanner.cpp|0|5|00.00%|[271, 272, 273, 274, 276]|

**Comment by JackWangCS** [comment]
> In addition to jdbc catalog, SR also supports the explicit creation of jdbc external tables to access data(ref:https://docs.starrocks.io/en-us/latest/using_starrocks/External_table#external-table-for-a-jdbc-compatible-database). We need to ensure that the TIME type also works in the latter way.
> 
> SR does not support the TIME type in the storage layer. If you specify a column of the TIME type when creating a table, an error will be reported.
> 
> However, the query of jdbc external table does not involve the storage layer. It may need special processing to ensure that the TIME type can be specified when creating the jdbc external table.
> 
> You can look at the logic related to table creation to ensure that what I said can work. @JackWangCS

@silverbullet233 Thanks for your comment, I will look at the logic related to the jdbc external tables and test whether it works.

**Comment by JackWangCS** [comment]
> > In addition to jdbc catalog, SR also supports the explicit creation of jdbc external tables to access data(ref:https://docs.starrocks.io/en-us/latest/using_starrocks/External_table#external-table-for-a-jdbc-compatible-database). We need to ensure that the TIME type also works in the latter way.
> > SR does not support the TIME type in the storage layer. If you specify a column of the TIME type when creating a table, an error will be reported.
> > However, the query of jdbc external table does not involve the storage layer. It may need special processing to ensure that the TIME type can be specified when creating the jdbc external table.
> > You can look at the logic related to table creation to ensure that what I said can work. @JackWangCS
> 
> @silverbullet233 Thanks for your comment, I will look at the logic related to the jdbc external tables and test whether it works.

Hi @silverbullet233 , just tested for jdbc external table case, it's supported. You can check the test log from this gist: https://gist.github.com/JackWangCS/fff11a9b66ea4a51cc261a1f8e40bb20.

**Comment by WholeWorld-Timothy** [comment]
Can you add some test cases？

**Review by zombee0 — APPROVED** [review]
LTTM

File: java-extensions/udf-extensions/src/main/java/com/starrocks/udf/UDFHelper.java [hunk 1]
```diff
@@ -192,6 +196,25 @@ private static void getDoubleBoxedResult(int numRows, Double[] boxedArr, long co
...
+            } else {
+                // Note: add the timezone offset back because Time#getTime() returns the GMT timestamp
+                dataArr[i] = (boxedArr[i].getTime() + timeZone.getRawOffset()) / 1000;
+            }
+        }
+
+        final long[] addrs = getAddrs(columnAddr);
+        // memcpy to uint8_t array
+        Platform.copyMemory(nulls, Platform.BYTE_ARRAY_OFFSET, null, addrs[0], numRows);
+        // memcpy to int array
```
Comment (stdpain) [inline java-extensions/udf-extensions/src/main/java/com/starrocks/udf/UDFHelper.java:214]:
```suggestion
        // memcpy to double array
```

**Review by stdpain — APPROVED** [review]

**Review by silverbullet233 — APPROVED** [review]

**Comment by andyziye** [comment]
https://github.com/Mergifyio backport branch-3.2

**Pull request merged**